From 52a48fcf8fa603a4ce9c7d4bad7ecc7edc93ab11 Mon Sep 17 00:00:00 2001
From: "Mikhail Tsukerman <mikhail.tsukerman@advantech.com>"
Date: Wed, 10 Jun 2026 18:08:15 +0000
Subject: [PATCH] Apply ECU150A1 board patch to current codebase

Upstream-Status: Inappropriate

- arch/arm/dts/Makefile: add ecu150a1.dtb entry
- board/freescale/imx8mq_evk/Makefile: update DDR timing objects for XPL_BUILD
- board/freescale/imx8mq_evk/imx8mq_evk.c: add PAD_CTL_PUE to UART_PAD_CTRL, add env vars
- board/freescale/imx8mq_evk/spl.c: change ddr_init to dram_timing_4g, add USB SDP check
- common/autoboot.c: add advupdate function, Enter-key only abort, call advupdate
- configs/imx8mq_evk_defconfig: apply config changes for ECU150A1
---
 board/freescale/imx8mq_evk/Makefile     |   6 +-
 board/freescale/imx8mq_evk/imx8mq_evk.c |   5 +-
 board/freescale/imx8mq_evk/spl.c        |  10 +-
 common/autoboot.c                       | 144 +++++++++++++++++++++++-
 configs/imx8mq_evk_defconfig            |  23 ++--
 6 files changed, 171 insertions(+), 18 deletions(-)

diff --git a/board/freescale/imx8mq_evk/Makefile b/board/freescale/imx8mq_evk/Makefile
index 63b7248b56a..b1983a62d78 100644
--- a/board/freescale/imx8mq_evk/Makefile
+++ b/board/freescale/imx8mq_evk/Makefile
@@ -8,9 +8,5 @@ obj-y += imx8mq_evk.o
 
 ifdef CONFIG_XPL_BUILD
 obj-y += spl.o
-ifdef CONFIG_IMX8MQ_4GB_DDR_TIMING
-obj-$(CONFIG_IMX8M_LPDDR4) += lpddr4_timing_4g.o
-else
-obj-$(CONFIG_IMX8M_LPDDR4) += lpddr4_timing_b0.o lpddr4_timing.o
-endif
+obj-$(CONFIG_IMX8M_LPDDR4) += lpddr4_timing_1g.o lpddr4_timing_2g.o lpddr4_timing_4g.o lpddr4_timing_b0.o
 endif
diff --git a/board/freescale/imx8mq_evk/imx8mq_evk.c b/board/freescale/imx8mq_evk/imx8mq_evk.c
index 9bd0869253d..707d1b23664 100644
--- a/board/freescale/imx8mq_evk/imx8mq_evk.c
+++ b/board/freescale/imx8mq_evk/imx8mq_evk.c
@@ -16,7 +16,7 @@
 #include <usb.h>
 #include <dwc3-uboot.h>
 
-#define UART_PAD_CTRL	(PAD_CTL_DSE6 | PAD_CTL_FSEL1)
+#define UART_PAD_CTRL	(PAD_CTL_DSE6 | PAD_CTL_FSEL1 | PAD_CTL_PUE)
 
 #define WDOG_PAD_CTRL	(PAD_CTL_DSE6 | PAD_CTL_HYS | PAD_CTL_PUE)
 
@@ -269,6 +269,9 @@ int board_late_init(void)
 	env_set("board_rev", "iMX8MQ");
 #endif
 
+	puts("DeviceName is ECU150A1.\n");
+	env_set("devicename", "ecu150a1");
+	env_set("bootargs_ram_capacity", "cma=384M");
 	return 0;
 }
 
diff --git a/board/freescale/imx8mq_evk/spl.c b/board/freescale/imx8mq_evk/spl.c
index c4f02655242..a0ebe323608 100644
--- a/board/freescale/imx8mq_evk/spl.c
+++ b/board/freescale/imx8mq_evk/spl.c
@@ -52,7 +52,7 @@ static void spl_dram_init(void)
 #else
 	/* ddr init */
 	if (soc_rev() >= CHIP_REV_2_1)
-		ddr_init(&dram_timing);
+		ddr_init(&dram_timing_4g);
 	else
 		ddr_init(&dram_timing_b0);
 #endif
@@ -233,6 +233,14 @@ void spl_board_init(void)
 			printf("\nsec_init failed!\n");
 	}
 
+#ifndef CONFIG_SPL_USB_SDP_SUPPORT
+	/* Serial download mode */
+	if (is_usb_boot()) {
+		puts("Back to ROM, SDP\n");
+		restore_boot_params();
+	}
+#endif
+
 	init_usb_clk();
 
 	puts("Normal Boot\n");
diff --git a/common/autoboot.c b/common/autoboot.c
index ec9f0fc4cbe..180b38dcfb9 100644
--- a/common/autoboot.c
+++ b/common/autoboot.c
@@ -44,6 +44,136 @@ DECLARE_GLOBAL_DATA_PTR;
 static int stored_bootdelay;
 static int menukey;
 
+static int advupdate(const char *ifname, const char *dev_part, const char *s)
+{
+	int ret = 0;
+	char buf[256] = {0};
+	bool haveconfig = true;
+
+	//start usb device
+	if (!strcmp(ifname,"usb"))
+	{
+		run_command("usb start", 0);
+	}
+
+	else if (!strcmp(ifname, "mmc"))
+	{
+		//load config file
+		env_set_ulong("filesize",0);
+		sprintf(buf, "fatload %s %s ${loadaddr} ${advufile};env import -t ${loadaddr} ${filesize}", ifname, dev_part);
+		run_command_list(buf, -1, 0);
+		if (env_get_ulong("filesize", 16, 0) == 0) {
+			memset(buf, 0, sizeof(buf));
+			sprintf(buf, "ext2load %s %s ${loadaddr} ${advufile};env import -t ${loadaddr} ${filesize}", ifname, dev_part);
+			run_command_list(buf, -1, 0);
+			if (env_get_ulong("filesize", 16, 0) == 0) {
+				// config file not found!
+				//printf("%s %s config file not found!\n", ifname, dev_part);
+				haveconfig = false;
+			}
+		}
+
+		// find update.bin file
+		if (!haveconfig)
+		{
+			memset(buf, 0, sizeof(buf));
+			sprintf(buf, "if test -e %s %s update.bin; then setenv advupdatebin y; fi", ifname, dev_part);
+			run_command_list(buf, -1, 0);
+			if (env_get_yesno("advupdatebin") != 1)
+			{
+				// update.bin file not found!
+				//printf("%s %s update bin file not found!\n", ifname, dev_part);
+				return 0;
+			}
+		}
+	}
+	else if (!strcmp(ifname, "nand"))
+	{
+	memset(buf, 0, sizeof(buf));
+	sprintf(buf, "ubi part NAND.recovery && ubifsmount ubi0;ubifsload ${loadaddr} ${advufile};env import -t ${loadaddr} ${filesize}");
+	run_command_list(buf, -1, 0);
+		run_command_list("ubifsumount", -1, 0);
+		if (env_get_ulong("filesize", 16, 0) == 0) {
+			// config file not found!
+			//printf("nand config file not found!!\n");
+			haveconfig = false;
+		}
+
+		// find update.bin file
+		if (!haveconfig)
+		{
+			memset(buf, 0, sizeof(buf));
+			sprintf(buf, "ubi part NAND.recovery && ubifsmount ubi0;ubifsload ${loadaddr} update.bin 0x10;env import -t ${loadaddr} 0x10");
+			run_command_list(buf, -1, 0);
+			run_command_list("ubifsumount", -1, 0);
+			if (env_get_ulong("filesize", 16, 0) == 0)
+			{
+				// update.bin file not found!
+				//printf("nand update bin file not found!!\n");
+				return 0;
+			}
+			else
+			{
+				run_command("setenv advupdatebin y", 0);
+			}
+		}
+	}
+
+	if (haveconfig)
+		printf("Config file found!\n");
+	else if (env_get_yesno("advupdatebin") == 1)
+		printf("update bin file found!\n");
+	else
+		return 0;
+
+	if ((env_get_ulong("filesize",16,0) > 0) || (env_get_yesno("advupdatebin") == 1))
+	{
+#if 1 //+= ramdisk boot
+		if ((env_get_yesno("advimage") == 1) || (env_get_yesno("advrootfs") == 1) || (env_get_yesno("advrecover") == 1) || (env_get_yesno("advupdatebin") == 1))
+		{
+			ret = 2;
+			printf("System recovery mode!\n");
+			if (!strcmp(ifname, "mmc"))
+			{
+				if (env_get_yesno("advupdatebin") == 1)
+				{
+					run_command("setenv mmcdev 0", 0);
+				}
+				else
+				{
+					if (!strcmp(dev_part, "1:1"))
+					{
+						run_command("setenv mmcdev 1", 0);
+					}
+					else if (!strcmp(dev_part, "0:3"))
+					{
+						run_command("setenv mmcdev 0", 0);
+					}
+				}
+				run_command("mmc dev ${mmcdev}", 0);
+
+				s = env_get("advrfs");
+				run_command_list(s, -1, 0);
+			}
+			else if (!strcmp(ifname, "usb"))
+			{
+				s = env_get("advrfs");
+				run_command_list(s, -1, 0);
+			}
+		}
+		else
+		{
+			printf("System boot mode!\n");
+			s = env_get("bootcmd");
+		}
+#endif
+	}
+	else
+		printf("Normal boot mode!\n");
+
+	return ret;
+}
+
 #if defined(CONFIG_AUTOBOOT_STOP_STR_CRYPT)
 #define AUTOBOOT_STOP_STR_CRYPT	CONFIG_AUTOBOOT_STOP_STR_CRYPT
 #else
@@ -385,9 +515,11 @@ static int abortboot_single_key(int bootdelay)
 	 * Check if key already pressed
 	 */
 	if (tstc()) {	/* we got a key press	*/
-		getchar();	/* consume input	*/
+		if (getchar() == 0x0d) // got enter key
+		{
 		puts("\b\b\b 0");
 		abort = 1;	/* don't auto boot	*/
+		}
 	}
 
 	while ((bootdelay > 0) && (!abort)) {
@@ -396,6 +528,8 @@ static int abortboot_single_key(int bootdelay)
 		ts = get_timer(0);
 		do {
 			if (tstc()) {	/* we got a key press	*/
+				if (getchar() == 0x0d) // got enter key
+				{
 				int key;
 
 				abort  = 1;	/* don't auto boot	*/
@@ -404,6 +538,7 @@ static int abortboot_single_key(int bootdelay)
 				if (IS_ENABLED(CONFIG_AUTOBOOT_USE_MENUKEY))
 					menukey = key;
 				break;
+				}
 			}
 			udelay(10000);
 		} while (!abort && get_timer(ts) < 1000);
@@ -527,6 +662,13 @@ void autoboot_command(const char *s)
 		if (lock)
 			prev = disable_ctrlc(1); /* disable Ctrl-C checking */
 
+		{
+			int ret = 0;
+			ret = advupdate("mmc", "1:1", s);
+			if (!ret)
+				ret = advupdate("mmc", "0:3", s);
+		}
+
 		run_command_list(s, -1, 0);
 
 		if (lock)
diff --git a/configs/imx8mq_evk_defconfig b/configs/imx8mq_evk_defconfig
index 42e35670124..78c277fd518 100644
--- a/configs/imx8mq_evk_defconfig
+++ b/configs/imx8mq_evk_defconfig
@@ -14,8 +14,8 @@ CONFIG_SYS_I2C_MXC_I2C1=y
 CONFIG_SYS_I2C_MXC_I2C2=y
 CONFIG_SYS_I2C_MXC_I2C3=y
 CONFIG_DM_GPIO=y
-CONFIG_DEFAULT_DEVICE_TREE="freescale/imx8mq-evk"
-CONFIG_USB_TCPC=y
+CONFIG_DEFAULT_DEVICE_TREE="freescale/ecu150a1"
+#CONFIG_USB_TCPC=y
 CONFIG_TARGET_IMX8MQ_EVK=y
 CONFIG_DM_RESET=y
 CONFIG_SPL_MMC=y
@@ -36,11 +36,12 @@ CONFIG_FIT_EXTERNAL_OFFSET=0x3000
 CONFIG_SPL_LOAD_FIT=y
 CONFIG_OF_SYSTEM_SETUP=y
 CONFIG_BOOTCOMMAND="run sr_ir_v2_cmd;bootflow scan -lb; run bsp_bootcmd"
-CONFIG_DEFAULT_FDT_FILE="imx8mq-evk.dtb"
+CONFIG_DEFAULT_FDT_FILE="ecu150a1.dtb"
 CONFIG_SYS_PBSIZE=1050
 CONFIG_BOARD_EARLY_INIT_F=y
 CONFIG_BOARD_LATE_INIT=y
-CONFIG_SPL_MAX_SIZE=0x26000
+#CONFIG_SPL_MAX_SIZE=0x26000
+CONFIG_SPL_MAX_SIZE=0x27000
 CONFIG_SPL_BOARD_INIT=y
 # CONFIG_SPL_SHARES_INIT_SP_ADDR is not set
 CONFIG_SPL_SYS_MALLOC=y
@@ -56,8 +57,9 @@ CONFIG_SYS_PROMPT="u-boot=> "
 CONFIG_SPL_USB_GADGET=y
 CONFIG_SPL_USB_SDP_SUPPORT=y
 # CONFIG_BOOTM_NETBSD is not set
-# CONFIG_CMD_EXPORTENV is not set
-# CONFIG_CMD_IMPORTENV is not set
+CONFIG_CMD_EXPORTENV=y
+CONFIG_CMD_IMPORTENV=y
+CONFIG_BOOTDELAY=1
 CONFIG_CMD_NVEDIT_EFI=y
 CONFIG_CMD_CRC32=y
 CONFIG_CRC32_VERIFY=y
@@ -183,12 +185,13 @@ CONFIG_EFI_VAR_BUF_SIZE=139264
 CONFIG_EFI_IGNORE_OSINDICATIONS=y
 CONFIG_EFI_CAPSULE_AUTHENTICATE=y
 CONFIG_EFI_CAPSULE_CRT_FILE="CRT.crt"
-CONFIG_OPTEE=y
-CONFIG_CMD_OPTEE_RPMB=y
+#CONFIG_OPTEE=y
+#CONFIG_CMD_OPTEE_RPMB=y
 CONFIG_EFI_MM_COMM_TEE=y
-CONFIG_TEE=y
+#CONFIG_TEE=y
 CONFIG_EFI_ESRT=y
 CONFIG_EFI_HAVE_CAPSULE_UPDATE=y
-CONFIG_FIT_SIGNATURE=y
+#CONFIG_FIT_SIGNATURE=y
+CONFIG_SMBIOS_PARSER=y
 
 CONFIG_BOOTDEV_ETH=n
-- 
2.43.0

